tsl/src/gapfill/exec.c

```diff
@@ -178,6 +189,182 @@ gapfill_state_create(CustomScan *cscan)
+{
+	Oid			result = InvalidOid;
+	HeapTuple	casttup;
+
+	casttup = SearchSysCache2(CASTSOURCETARGET, ObjectIdGetDatum(source), ObjectIdGetDatum(target));
+	if (HeapTupleIsValid(casttup))
+	{
+		Form_pg_cast castform = (Form_pg_cast) GETSTRUCT(casttup);
+
+		result = castform->castfunc;
+		ReleaseSysCache(casttup);
+	}
+
+	if (!OidIsValid(result))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("error looking up cast from %d to %d", source, target)));
+
+	return result;
+}
+
+/*
+ * returns true if v1 and v2 reference the same object
+ */
+static bool
+var_equal(Var *v1, Var *v2)
+{
+	return v1->varno == v2->varno && v1->varattno == v2->varattno && v1->vartype == v2->vartype;
+}
+
+static void
+deduce_gapfill_boundary(GapFillState *state, GapFillBoundary boundary, int64 *value)
+{
+	CustomScan *cscan = castNode(CustomScan, state->csstate.ss.ps.plan);
+	FuncExpr   *func = linitial(cscan->custom_private);
+	FromExpr   *jt = lfourth(cscan->custom_private);
+	List	   *vars;
+	ListCell   *lc;
+
+	const char *op1,
+			   *op2;
+
+	if (GAPFILL_START == boundary)
+	{
+		op1 = ">";
+		op2 = ">=";
+	}
+	else
+	{
+		op1 = "<";
+		op2 = "<=";
+	}
+
+	/*
+	 * if the second argument to time_bucket_gapfill is not a column reference
+	 * we cannot match WHERE clause to the time column, we don't check for Var
+	 * directly though cause it might be wrapped in a cast
+	 */
+	vars = pull_var_clause(lsecond(func->args), 0);
+	if (list_length(vars) != 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid ts: need to refer to column if no start or finish is supplied")));
+
+	foreach(lc, (List *) jt->quals)
+	{
+		OpExpr	   *opexpr = lfirst(lc);
+
+		/*
+		 * check if left side of OpExpr is Var and matches Var passed in as
+		 * 2nd arg to time_bucket_gapfill call
+		 */
+		if (IsA(lfirst(lc), OpExpr) &&IsA(linitial(opexpr->args), Var) &&var_equal(linitial(vars), linitial(opexpr->args)))
+		{
+			char	   *oprname = get_oprname(opexpr->opno);
+
+			if (strcmp(oprname, op1) == 0 || strcmp(oprname, op2) == 0)
+			{
+				Datum		arg_value;
+				bool		isnull;
+				Expr	   *expr = lsecond(opexpr->args);
+
+				/*
+				 * add an explicit cast here to the type we require if
+				 * necessary
+				 */
+				if (exprType(lsecond(opexpr->args)) != state->gapfill_typid)
+				{
+					Oid			cast_oid = get_castfunc(exprType(lsecond(opexpr->args)), state->gapfill_typid);
+
+					expr = (Expr *) makeFuncExpr(cast_oid, state->gapfill_typid, list_make1(expr), InvalidOid, InvalidOid, 0);
+				}
+
+				arg_value = gapfill_exec_expr(state, expr, &isnull);
+
+				/*
+				 * if the expression evaluates to NULL we keep looking
+				 */
+				if (isnull)
+					continue;
+
+				/*
+				 * for gapfill start we pass the value through time_bucket to
+				 * align with chunk boundary
+				 */
+				if (GAPFILL_START == boundary)
+				{
+					int64		value = gapfill_datum_get_internal(arg_value, state->gapfill_typid);
```

Comment:
Different name for `value` argument given that value is already scoped in broader function?  While correct, could lead to programmer confusion.

Reply:
fixed